src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -1085,47 +1102,139 @@ private static final class SimpleStringBuilderStrategy {
+            SoftReference<MethodHandlePair> weakConstructorHandle = CACHE.get(erasedArgs);
+            if (weakConstructorHandle != null) {
+                MethodHandlePair handlePair = weakConstructorHandle.get();
+                if (handlePair != null) {
+                    try {
+                        var instance = handlePair.constructor.invoke(constants);
+                        return handlePair.concatenator.bindTo(instance);
+                    } catch (Throwable e) {
+                        throw new StringConcatException("Exception while utilizing the hidden class", e);
+                    }
+                }
+            }
+            ClassDesc concatClass = ConstantUtils.binaryNameToDesc(className);
+            byte[] classBytes = ClassFile.of().build(concatClass,
                     new Consumer<ClassBuilder>() {
                         @Override
                         public void accept(ClassBuilder clb) {
                             clb.withFlags(ClassFile.ACC_FINAL | ClassFile.ACC_SUPER | ClassFile.ACC_SYNTHETIC)
+                                .withField(LENGTH, CD_int, FINAL_STABLE_FIELD)
+                                .withField(CODER, CD_byte, FINAL_STABLE_FIELD)
+                                .withField(CONSTANTS, CD_Array_String, FINAL_STABLE_FIELD)
+                                .withMethodBody("<init>",
+                                        ARRAY_STRING_TO_VOID,
+                                        ClassFile.ACC_PRIVATE,
+                                        generateConstructor(concatClass, constants.length)
+                                )
                                 .withMethodBody(METHOD_NAME,
-                                        ConstantUtils.methodTypeDesc(erasedArgs),
-                                        ClassFile.ACC_FINAL | ClassFile.ACC_PRIVATE | ClassFile.ACC_STATIC,
-                                        generateMethod(constants, erasedArgs));
+                                        ConstantUtils.methodTypeDesc(concatArgs),
+                                        ClassFile.ACC_FINAL | ClassFile.ACC_PRIVATE,
+                                        generateConcatMethod(concatClass, concatArgs));
                     }});
             try {
                 var hiddenClass = lookup.makeHiddenClassDefiner(className, classBytes, Set.of(), DUMPER)
                         .defineClass(true, null);
-                return lookup.findStatic(hiddenClass, METHOD_NAME, erasedArgs);
-            } catch (Exception e) {
+
+                MethodHandle constructorHandle = lookup.findConstructor(hiddenClass, MethodType.methodType(void.class, String[].class));
+                var instance = hiddenClass.cast(constructorHandle.invoke(constants));
+                MethodHandle handle = lookup.findVirtual(hiddenClass, METHOD_NAME, concatArgs);
+                CACHE.put(concatArgs, new SoftReference<>(new MethodHandlePair(constructorHandle, handle)));
```

Comment:
It seems possible that the MethodHandlePair gets GC'd while the hidden class is still reachable. Will we spin a new hidden class in that case?

Reply:
Yes, while imperfect this should work, and the alternatives seem worse.
- WeakReference -> No-one actually strongly references the `MethodHandlePair`, so it'll be quickly GC'd
- Strong reference -> The `MethodHandlePair` now creates a strong reference to the `MethodType`, so now we'll never be dereferenced. 

 Using a SoftReference is in principle not any different from how the `LambdaForm.transformCache` operates today.

Reply:
Sorry for not seeing and replying earlier, but I think we can link to the pair from the hidden class, such as in a static field; this ensures MethodHandlePair lives as long as the hidden class at least, and the MHPair can be GC'd with the implementation class gone.